File: lib/disco_manager.py
import time
import board
import neopixel
import math
import asyncio
import random

class DiscoMode:
    def __init__(self, sensor_manager, pixel_pin=board.NEOPIXEL, num_pixels=1, brightness=1.0):
        self.sensor_manager = sensor_manager
        self.pixel_pin = pixel_pin
        self.num_pixels = num_pixels

        # Brightness handling
        self.base_brightness = 0.3
        self.wild_brightness = 1.0
        self.brightness = self.base_brightness
        self._last_shake_time = 0
        self._shake_timeout = 5  # seconds to return to dim after shake

        self.pixels = neopixel.NeoPixel(
            self.pixel_pin,
            self.num_pixels,
            brightness=1.0,  # Hardware brightness stays at max
            auto_write=False
        )
        self.is_active = False
        self.task = None
        self.loop = asyncio.get_event_loop()

    def _apply_brightness(self, r, g, b):
        """Apply software brightness scaling."""
        return (
            int(r * self.brightness),
            int(g * self.brightness),
            int(b * self.brightness)
        )

    def set_green(self):
        """Set LED to green."""
        self._stop_disco()
        self.pixels.fill(self._apply_brightness(0, 255, 0))
        self.pixels.show()

    def set_red(self):
        """Set LED to red."""
        self._stop_disco()
        self.pixels.fill(self._apply_brightness(255, 0, 0))
        self.pixels.show()

    def turn_off(self):
        """Turn off the LED."""
        self._stop_disco()
        self.pixels.fill((0, 0, 0))
        self.pixels.show()

    def _stop_disco(self):
        if self.is_active:
            self.is_active = False
            if self.task is not None:
                self.task.cancel()
                self.task = None

    def start_disco(self):
        """Start the disco (rainbow) effect."""
        if not self.is_active:
            self.is_active = True
            self.brightness = self.base_brightness  # Reset to dim every time it's started
            self.wild_latched = False               # Reset latch
            self._last_shake_time = 0               # Clear shake timestamp
            print("Disco Mode ON (starting dim)")
            self.task = self.loop.create_task(self._run_disco_effect())


    async def _run_disco_effect(self):
        """Run the rainbow effect with latched wild mode based on gravity."""
        base_sleep = 0.2
        wild_sleep = 0.1
        color_step_base = 20

        j = 0
        self.wild_latched = False

        while self.is_active:
            grav = self.sensor_manager.get_grav()
            gx, gy, gz = grav if grav else (0, 0, 0)

            now = time.monotonic()

            # Detect shake
            if abs(gx) > 11 or abs(gy) > 11 or abs(gz) > 11:
                self.wild_latched = True
                self._last_shake_time = now
                self.brightness = self.wild_brightness
                print("🔥 Wild mode latched ON! Brightness increased.")

            # Return to dim after timeout
            if self.wild_latched and now - self._last_shake_time > self._shake_timeout:
                self.brightness = self.base_brightness
                print("🌙 Calm mode. Brightness reduced.")

            is_wild = self.brightness == self.wild_brightness
            sleep_time = wild_sleep if is_wild else base_sleep

            # Choose color
            if is_wild:
                hue = random.random()
            else:
                hue = (j % 360) / 360.0
                j += color_step_base

            # Apply color
            color = self.hsv_to_rgb(hue, 1.0, 1.0)
            color = self._apply_brightness(*color)
            self.pixels.fill(color)
            self.pixels.show()

            await asyncio.sleep(sleep_time)

        # Cleanup
        self.pixels.fill((0, 0, 0))
        self.pixels.show()
        self.wild_latched = False

    def hsv_to_rgb(self, h, s, v):
        """Convert HSV to RGB."""
        i = int(h * 6)
        f = h * 6 - i
        p = v * (1 - s)
        q = v * (1 - f * s)
        t = v * (1 - (1 - f) * s)
        i = i % 6
        if i == 0: return int(v * 255), int(t * 255), int(p * 255)
        if i == 1: return int(q * 255), int(v * 255), int(p * 255)
        if i == 2: return int(p * 255), int(v * 255), int(t * 255)
        if i == 3: return int(p * 255), int(q * 255), int(v * 255)
        if i == 4: return int(t * 255), int(p * 255), int(v * 255)
        if i == 5: return int(v * 255), int(p * 255), int(q * 255)

File: Main board C++/src/disco_manager.cpp
#include "disco_manager.h"
#include <math.h>

// ── Lifecycle ────────────────────────────────────────────────────────

void DiscoManager::begin() {
    pixel_.begin();
    pixel_.setBrightness(255);   // HW brightness max; we software-scale
    pixel_.show();
}

// ── Main update — call every loop() iteration ───────────────────────

void DiscoManager::update(float accelX, float accelY, float accelZ) {
    uint32_t now = millis();

    if (effect_ == DiscoEffect::DISCO) {
        // --- Shake detection for wild mode ---
        if (fabsf(accelX) > Disco::SHAKE_THRESHOLD ||
            fabsf(accelY) > Disco::SHAKE_THRESHOLD ||
            fabsf(accelZ) > Disco::SHAKE_THRESHOLD) {
            wildLatched_ = true;
            lastShakeTime_ = now;
            brightness_ = Disco::WILD_BRIGHTNESS;
        }

        // Return to dim after timeout
        if (wildLatched_ && (now - lastShakeTime_ > Disco::SHAKE_TIMEOUT_MS)) {
            wildLatched_ = false;
            brightness_ = Disco::BASE_BRIGHTNESS;
        }

        bool isWild = (brightness_ == Disco::WILD_BRIGHTNESS);
        uint32_t interval = isWild ? Disco::WILD_SLEEP_MS : Disco::BASE_SLEEP_MS;

        if (now - lastFrameTime_ >= interval) {
            lastFrameTime_ = now;

            float hue;
            if (isWild) {
                // Random hue in wild mode
                hue = random(360) / 360.0f;
            } else {
                hue = hueStep_ / 360.0f;
                hueStep_ = (hueStep_ + Disco::COLOR_STEP_BASE) % 360;
            }

            uint8_t r, g, b;
            hsvToRgb(hue, 1.0f, 1.0f, r, g, b);
            setPixel(r, g, b);
        }

    } else if (effect_ == DiscoEffect::PURPLE_PULSE) {
        // ~33 Hz update (30 ms)
        if (now - lastPulseTime_ >= 30) {
            lastPulseTime_ = now;

            float pulseBright = 0.55f + 0.45f * sinf(pulsePhase_);
            pulsePhase_ += 0.3f;

            uint8_t r = (uint8_t)(255.0f * pulseBright);
            uint8_t g = 0;
            uint8_t b = (uint8_t)(255.0f * pulseBright);
            pixel_.setPixelColor(0, pixel_.Color(r, g, b));
            pixel_.show();
        }
    }
    // STATIC and OFF need no per-frame work
}

// ── Static colours ──────────────────────────────────────────────────

void DiscoManager::setRed() {
    stopAll();
    effect_ = DiscoEffect::STATIC;
    setPixel(255, 0, 0);
}

void DiscoManager::setGreen() {
    stopAll();
    effect_ = DiscoEffect::STATIC;
    setPixel(0, 255, 0);
}

void DiscoManager::setBlue() {
    stopAll();
    effect_ = DiscoEffect::STATIC;
    setPixel(0, 0, 255);
}

void DiscoManager::setWhite() {
    stopAll();
    effect_ = DiscoEffect::STATIC;
    setPixel(255, 255, 255);
}

void DiscoManager::turnOff() {
    stopAll();
}

// ── Animated effects ────────────────────────────────────────────────

void DiscoManager::startDisco() {
    if (effect_ == DiscoEffect::DISCO) return;   // already running
    stopAll();
    effect_        = DiscoEffect::DISCO;
    brightness_    = Disco::BASE_BRIGHTNESS;
    wildLatched_   = false;
    lastShakeTime_ = 0;
    hueStep_       = 0;
    lastFrameTime_ = millis();
}

void DiscoManager::setPurple() {
    stopAll();
    effect_        = DiscoEffect::PURPLE_PULSE;
    pulsePhase_    = 0.0f;
    lastPulseTime_ = millis();
}

// ── Internal helpers ────────────────────────────────────────────────

void DiscoManager::stopAll() {
    effect_ = DiscoEffect::OFF;
    wildLatched_ = false;
    brightness_ = Disco::BASE_BRIGHTNESS;
    pixel_.setPixelColor(0, 0);
    pixel_.show();
}

void DiscoManager::setPixel(uint8_t r, uint8_t g, uint8_t b) {
    // Apply software brightness scaling
    uint8_t rr = (uint8_t)(r * brightness_);
    uint8_t gg = (uint8_t)(g * brightness_);
    uint8_t bb = (uint8_t)(b * brightness_);
    pixel_.setPixelColor(0, pixel_.Color(rr, gg, bb));
    pixel_.show();
}

void DiscoManager::hsvToRgb(float h, float s, float v,
                             uint8_t& r, uint8_t& g, uint8_t& b) {
    // Identical algorithm to Python's DiscoMode.hsv_to_rgb
    int i = (int)(h * 6.0f);
    float f = h * 6.0f - i;
    float p = v * (1.0f - s);
    float q = v * (1.0f - f * s);
    float t = v * (1.0f - (1.0f - f) * s);
    i = i % 6;
    float rv, gv, bv;
    switch (i) {
        case 0: rv = v; gv = t; bv = p; break;
        case 1: rv = q; gv = v; bv = p; break;
        case 2: rv = p; gv = v; bv = t; break;
        case 3: rv = p; gv = q; bv = v; break;
        case 4: rv = t; gv = p; bv = v; break;
        default: rv = v; gv = p; bv = q; break;   // case 5
    }
    r = (uint8_t)(rv * 255.0f);
    g = (uint8_t)(gv * 255.0f);
    b = (uint8_t)(bv * 255.0f);
}
